File: clients/RookKit/Sources/RookKit/Models/ApiTypes.swift
import Foundation

public struct AgentDefinition: Codable, Equatable, Identifiable {
    public let id: String
    public let parentId: String?

    public init(id: String, parentId: String?) {
        self.id = id
        self.parentId = parentId
    }
}

/// Wraps the raw session record JSON so resume can send the record back to
/// `POST /api/agent/start` verbatim, including fields this app doesn't model.
public struct AgentSessionSummary: Equatable, Identifiable {
    public let raw: JSONValue

    public init(raw: JSONValue) {
        self.raw = raw
    }

    public var id: String { raw["id"]?.stringValue ?? "" }
    public var agent: String { raw["agent"]?.stringValue ?? "" }
    public var name: String { raw["name"]?.stringValue ?? "default" }
    public var running: Bool { raw["running"]?.boolValue ?? false }
    public var connectedClients: Int { Int(raw["connectedClients"]?.numberValue ?? 0) }

    public var createdAt: Date? {
        guard let iso = raw["createdAt"]?.stringValue else {
            return nil
        }
        let withFraction = ISO8601DateFormatter()
        withFraction.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = withFraction.date(from: iso) {
            return date
        }
        return ISO8601DateFormatter().date(from: iso)
    }

    public var createdAtLabel: String {
        guard let date = createdAt else {
            return ""
        }
        let formatter = DateFormatter()
        formatter.dateStyle = .medium
        formatter.timeStyle = .short
        return formatter.string(from: date)
    }
}

public struct EnvironmentArtifactPreview: Codable, Equatable, Identifiable {
    public let id: String
    public let files: [String: String]

    public init(id: String, files: [String: String]) {
        self.id = id
        self.files = files
    }

    public var sortedFilePaths: [String] {
        files.keys.sorted()
    }
}

/// Phone -> server payload asking which `loc:` environments are likely
/// available at the current location (issue #42, phase 1).
public struct IdentifyAvailableRequest: Encodable, Equatable {
    public var latitude: Double
    public var longitude: Double
    public var horizontalAccuracy: Double?
    public var source: String?
    public var dwellSeconds: Double?
    public var isStationary: Bool?
    public var speedMetersPerSecond: Double?
    public var observedAt: String?

    public init(
        latitude: Double,
        longitude: Double,
        horizontalAccuracy: Double? = nil,
        source: String? = nil,
        dwellSeconds: Double? = nil,
        isStationary: Bool? = nil,
        speedMetersPerSecond: Double? = nil,
        observedAt: String? = nil
    ) {
        self.latitude = latitude
        self.longitude = longitude
        self.horizontalAccuracy = horizontalAccuracy
        self.source = source
        self.dwellSeconds = dwellSeconds
        self.isStationary = isStationary
        self.speedMetersPerSecond = speedMetersPerSecond
        self.observedAt = observedAt
    }
}

/// A ranked candidate environment returned by `identify` / `register-location`.
public struct EnvironmentCandidate: Codable, Equatable, Identifiable {
    public let environmentId: String
    public let displayName: String
    public let operator_: String?
    public let storeNumber: String?
    public let address: String?
    public let latitude: Double?
    public let longitude: Double?
    public let website: String?
    public let distanceMeters: Double?
    public let confidence: Double
    public let matchReasons: [String]
    public let hasKnownEnvironment: Bool
    public let possibleSkills: [String]?

    public var id: String { environmentId }

    enum CodingKeys: String, CodingKey {
        case environmentId, displayName
        case operator_ = "operator"
        case storeNumber, address, latitude, longitude, website, distanceMeters, confidence, matchReasons, hasKnownEnvironment, possibleSkills
    }

    public init(
        environmentId: String,
        displayName: String,
        operator_: String?,
        storeNumber: String?,
        address: String?,
        latitude: Double?,
        longitude: Double?,
        website: String?,
        distanceMeters: Double?,
        confidence: Double,
        matchReasons: [String],
        hasKnownEnvironment: Bool,
        possibleSkills: [String]?
    ) {
        self.environmentId = environmentId
        self.displayName = displayName
        self.operator_ = operator_
        self.storeNumber = storeNumber
        self.address = address
        self.latitude = latitude
        self.longitude = longitude
        self.website = website
        self.distanceMeters = distanceMeters
        self.confidence = confidence
        self.matchReasons = matchReasons
        self.hasKnownEnvironment = hasKnownEnvironment
        self.possibleSkills = possibleSkills
    }
}

public struct RepositoryReadError: Codable, Equatable, Identifiable {
    public let code: String
    public let message: String
    public let repository: String
    public let environmentId: String
    public let bundleId: String?
    public let path: String?

    public init(code: String, message: String, repository: String, environmentId: String, bundleId: String?, path: String?) {
        self.code = code
        self.message = message
        self.repository = repository
        self.environmentId = environmentId
        self.bundleId = bundleId
        self.path = path
    }

    public var id: String { [code, repository, environmentId, bundleId ?? "", path ?? ""].joined(separator: "|") }
}

public struct EnvironmentBundlePreview: Codable, Equatable, Identifiable {
    public let id: String
    public let bundleId: String
    public let environmentId: String
    public let repository: String
    public let valid: Bool
    public let bundleHash: String
    public let skills: [EnvironmentArtifactPreview]
    public let mcpServers: [EnvironmentArtifactPreview]
    public let apps: [EnvironmentArtifactPreview]
    public let errors: [RepositoryReadError]

    public init(id: String, bundleId: String, environmentId: String, repository: String, valid: Bool, bundleHash: String, skills: [EnvironmentArtifactPreview], mcpServers: [EnvironmentArtifactPreview], apps: [EnvironmentArtifactPreview], errors: [RepositoryReadError]) {
        self.id = id
        self.bundleId = bundleId
        self.environmentId = environmentId
        self.repository = repository
        self.valid = valid
        self.bundleHash = bundleHash
        self.skills = skills
        self.mcpServers = mcpServers
        self.apps = apps
        self.errors = errors
    }

    public var allArtifacts: [EnvironmentArtifactPreview] {
        skills + mcpServers + apps
    }

    public var allFilePaths: [String] {
        allArtifacts.flatMap(\.sortedFilePaths).sorted()
    }

    public func content(for path: String) -> String? {
        for artifact in allArtifacts {
            if let content = artifact.files[path] {
                return content
            }
        }
        return nil
    }
}

public struct EnvironmentPreview: Codable, Equatable {
    public let environmentId: String
    public let bundles: [EnvironmentBundlePreview]

    public init(environmentId: String, bundles: [EnvironmentBundlePreview]) {
        self.environmentId = environmentId
        self.bundles = bundles
    }
}

public struct EnvironmentOffer: Equatable {
    public let environmentId: String
    public let bundleId: String
    public let bundleHash: String
    public let sourceName: String?
    public let canonicalSourceUrl: String?
    public let skills: [String]
    public let mcpServers: [String]
    public let apps: [String]

    public init(environmentId: String, bundleId: String, bundleHash: String, sourceName: String?, canonicalSourceUrl: String?, skills: [String], mcpServers: [String], apps: [String]) {
        self.environmentId = environmentId
        self.bundleId = bundleId
        self.bundleHash = bundleHash
        self.sourceName = sourceName
        self.canonicalSourceUrl = canonicalSourceUrl
        self.skills = skills
        self.mcpServers = mcpServers
        self.apps = apps
    }
}

File: clients/android/app/src/main/java/com/rookery/rook/model/ApiTypes.kt
// Mirrors clients/RookKit/Sources/RookKit/Models/ApiTypes.swift
package com.rookery.rook.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException
import java.time.format.FormatStyle

@Serializable
data class AgentDefinition(
    val id: String,
    val parentId: String?
)

data class AgentSessionSummary(val raw: JsonObject) {
    val id: String get() = raw["id"]?.stringValue ?: raw["sessionId"]?.stringValue ?: ""
    val agent: String get() = raw["agent"]?.stringValue ?: raw["_meta"]?.jsonObject?.get("runtimeId")?.stringValue ?: ""
    val name: String get() = raw["name"]?.stringValue ?: raw["title"]?.stringValue ?: "default"
    val running: Boolean get() = raw["running"]?.boolValue ?: false
    val connectedClients: Int get() = (raw["connectedClients"]?.numberValue ?: 0.0).toInt()

    val createdAt: Instant?
        get() = parseInstant(raw["createdAt"]?.stringValue ?: raw["_meta"]?.jsonObject?.get("startedAt")?.stringValue)

    val updatedAt: Instant?
        get() = parseInstant(raw["updatedAt"]?.stringValue)

    val createdAtLabel: String get() = formatInstant(createdAt)
    val updatedAtLabel: String get() = formatInstant(updatedAt)

    private fun parseInstant(iso: String?): Instant? {
        iso ?: return null
        return try { Instant.parse(iso) } catch (_: DateTimeParseException) { null }
    }

    private fun formatInstant(instant: Instant?): String {
        val date = instant ?: return ""
        val formatter = DateTimeFormatter.ofLocalizedDateTime(FormatStyle.MEDIUM, FormatStyle.SHORT).withZone(ZoneId.systemDefault())
        return formatter.format(date)
    }
}

@Serializable
data class EnvironmentArtifactPreview(
    val id: String,
    val files: Map<String, String>
) {
    val sortedFilePaths: List<String> get() = files.keys.sorted()
}

// Phone -> server payload asking which location: environments are likely available at the
// current location (issue #42, phase 1).
@Serializable
data class IdentifyAvailableRequest(
    val latitude: Double,
    val longitude: Double,
    val horizontalAccuracy: Double? = null,
    val source: String? = null,
    val dwellSeconds: Double? = null,
    val isStationary: Boolean? = null,
    val speedMetersPerSecond: Double? = null,
    val observedAt: String? = null
)

// A ranked candidate environment returned by identify-available.
@Serializable
data class EnvironmentCandidate(
    val environmentId: String,
    val displayName: String,
    @SerialName("operator") val operator_: String?,
    val storeNumber: String?,
    val address: String?,
    val latitude: Double?,
    val longitude: Double?,
    val website: String?,
    val distanceMeters: Double?,
    val confidence: Double,
    val matchReasons: List<String>,
    val hasKnownEnvironment: Boolean,
    val possibleSkills: List<String>?
) {
    val id: String get() = environmentId
}

@Serializable
data class RepositoryReadError(
    val code: String,
    val message: String,
    val repository: String,
    val environmentId: String,
    val bundleId: String?,
    val path: String?
) {
    val id: String get() = listOf(code, repository, environmentId, bundleId ?: "", path ?: "").joinToString("|")
}

@Serializable
data class EnvironmentBundlePreview(
    val id: String,
    val bundleId: String,
    val environmentId: String,
    val repository: String,
    val valid: Boolean,
    val bundleHash: String = "",
    val skills: List<EnvironmentArtifactPreview>,
    val mcpServers: List<EnvironmentArtifactPreview>,
    val apps: List<EnvironmentArtifactPreview>,
    val errors: List<RepositoryReadError>
) {
    val allArtifacts: List<EnvironmentArtifactPreview> get() = skills + mcpServers + apps
    val allFilePaths: List<String> get() = allArtifacts.flatMap { it.sortedFilePaths }.sorted()

    fun content(path: String): String? {
        for (artifact in allArtifacts) {
            artifact.files[path]?.let { return it }
        }
        return null
    }
}

@Serializable
data class EnvironmentPreview(
    val environmentId: String,
    val bundles: List<EnvironmentBundlePreview>
)

@Serializable
data class CandidateEnvironmentRecord(
    val id: String,
    val metadata: JsonObject
)

data class EnvironmentOffer(
    val environmentId: String,
    val displayName: String?,
    val bundleId: String,
    val bundleHash: String,
    val sourceName: String?,
    val canonicalSourceUrl: String?,
    val skills: List<String>,
    val mcpServers: List<String>,
    val apps: List<String>
)

// One environment row from GET /api/environments/list.
@Serializable
data class EnvironmentListItem(
    val environmentId: String,
    val displayName: String,
    val sourceName: String? = null,
    val status: String,
    val lastTouchedAt: String,
    val entered: Boolean,
    val bundleCount: Int,
    val approvedBundleCount: Int
) {
    val id: String get() = environmentId
}
